sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,348 @@
+	}
+
+	async deleteMoveFunctionDefinition(ref: { package: string; module: string; function: string }) {
+		const functionName = `${normalizeSuiAddress(ref.package)}::${ref.module}::${ref.function}`;
+		await this.delete('MoveFunction', functionName);
+	}
+}
+
+export class InMemoryCache extends AsyncCache {
+	#caches = {
+		OwnedObject: new Map<string, ObjectCacheEntry>(),
+		SharedOrImmutableObject: new Map<string, ObjectCacheEntry>(),
+		MoveFunction: new Map<string, MoveFunctionCacheEntry>(),
+	};
+
+	protected async get<T extends keyof CacheEntryTypes>(type: T, key: string) {
+		return (this.#caches[type].get(key) as CacheEntryTypes[T]) ?? null;
+	}
+
+	protected async set<T extends keyof CacheEntryTypes>(
+		type: T,
+		key: string,
+		value: CacheEntryTypes[T],
+	) {
+		(this.#caches[type] as Map<string, typeof value>).set(key, value as never);
+	}
+
+	protected async delete<T extends keyof CacheEntryTypes>(type: T, key: string) {
+		this.#caches[type].delete(key);
+	}
+
+	async clear<T extends keyof CacheEntryTypes>(type?: T) {
+		if (type) {
+			this.#caches[type].clear();
+		} else {
+			for (const cache of Object.values(this.#caches)) {
+				cache.clear();
+			}
+		}
+	}
+}
+
+interface ObjectCacheOptions {
+	cache?: AsyncCache;
+	address: string;
+}
+
+export class ObjectCache {
+	#cache: AsyncCache;
+	#address: string;
+
+	constructor({ cache = new InMemoryCache(), address }: ObjectCacheOptions) {
+		this.#cache = cache;
+		this.#address = normalizeSuiAddress(address);
+	}
+
+	asPlugin(): TransactionBlockPlugin {
+		return async (blockData, _options, next) => {
+			const unresolvedObjects = blockData.inputs
+				.filter((input) => input.UnresolvedObject)
+				.map((input) => input.UnresolvedObject!.objectId);
+
+			const cached = (await this.#cache.getObjects(unresolvedObjects)).filter(
+				(obj) => obj !== null,
+			);
+
+			const byId = new Map(cached.map((obj) => [obj!.objectId, obj]));
+
+			for (const input of blockData.inputs) {
+				if (!input.UnresolvedObject) {
+					continue;
+				}
+
+				const cached = byId.get(input.UnresolvedObject.objectId);
+
+				if (!cached) {
+					continue;
+				}
+
+				if (cached.initialSharedVersion && !input.UnresolvedObject.initialSharedVersion) {
+					input.UnresolvedObject.initialSharedVersion = cached.initialSharedVersion;
+				}
+
+				if (cached.version && !input.UnresolvedObject.version) {
+					input.UnresolvedObject.version = cached.version;
+				}
+
+				if (cached.digest && !input.UnresolvedObject.digest) {
+					input.UnresolvedObject.digest = cached.digest;
+				}
+			}
+
+			await Promise.all(
+				blockData.transactions.map(async (tx) => {
+					if (tx.MoveCall) {
+						const def = await this.getMoveFunctionDefinition({
+							package: tx.MoveCall.package,
+							module: tx.MoveCall.module,
+							function: tx.MoveCall.function,
+						});
+
+						if (def) {
+							tx.MoveCall.argumentTypes = def.parameters;
+						}
+					}
+				}),
+			);
+
+			await next();
```

Comment:
we aren't caching objects after resolution yet, just using the effects, but it's easy enough to do if we want that